content/blog/2024/2024-05-25-creating-an-npm-package .md

```diff
@@ -0,0 +1,534 @@
+- Increment the `MINOR` number for added functionality in a backward-compatible manner.
+- Increment the `MAJOR` number for significant changes or incompatible API changes.
+
+These versions should be updated appropriately whenever the NPM package is modified or changed, the same package version number cannot be published twice on the NPM registry. 
+
+Next, we will update our `package.json` file with important properties and script commands. 
+
+Here's what our NPM package.json should look like now:
+```json 
+{
+  "name": "validate-npm-pc",
+  "version": "1.0.0",
+  "description": "A comprehensive library for validating user inputs including emails, mobile numbers, and social media links.",
+  "main": "dist/cjs/index.js",
+  "module": "dist/esm/index.js",
+  "types": "dist/types/index.d.ts",
+  "files": [
+    "/dist"
+  ],
+  "scripts": {
+    "build": "tsc --project tsconfig.json && tsc --project tsconfig.cjs.json",
+    "release": "npm run build && changeset publish",
+    "test": "jest"
+  },
+  "keywords": [
+    "validate"
+  ],
+  "author": "ajibadde",
+  "license": "ISC",
+  "devDependencies": {
+    "@changesets/cli": "^2.27.1",
+    "@types/jest": "^29.5.12",
+    "jest": "^29.7.0",
+    "ts-jest": "^29.1.2",
+    "typescript": "^5.4.5"
+  }
+}
+```
+Our `package.json` now has these additional fields:
+- **main**: This specifies the entry point for CommonJS users. When someone uses `require("validate-npm-pc"),` Node.js will look for `dist/cjs/index.js`.
+- **module**: This specifies the entry point for ES module users. When someone uses `import from "validate-npm-pc"`, tools like Webpack or Rollup will look for `dist/esm/index.js`.
+- **types**: This specifies the location of the TypeScript declaration file. This helps TypeScript understand the types when our package is used.
+- **files**: An array specifying which files should be included when our package is published.
+- **keywords**: An array of keywords to enhance our package's searchability on the NPM website.
+- **scripts**: Define commands for building, releasing, and testing the project.
+    - **build**: This script compiles TypeScript files into JavaScript using the `tsc`. It runs two separate builds using the typescript configuration files we will be creating (`tsconfig.json` and `tsconfig.cjs.json`).
+    - **release**: This script first runs the `build` script, and if successful, it then publishes the changes using `changeset publish`.
+    - **test**: This runs our project tests using Jest
+
+Next, update the `.gitignore` to exclude unnecessary files from being included in our GitHub repository
+
+To do this copy and paste the following into the `.gitignore` file:
+```
+dist
+node_modules
+```
+
+# Step 2: Initializing Helper Package
+Before proceeding with our package development, we need to initialize the necessary helper dependencies for our NPM package. We'll be setting up TypeScript, Jest, and Changesets in our project.
+
+## Initialize Typescript
```

Comment:
Some headers were using the "-ing" form and some not. I changed it so that all headers use the "-ing" form.

Reply:
Okay Thank You